sftp/sftp.go

```diff
@@ -0,0 +1,228 @@
+	// ssh.ServerConfig so we had to implement it separately.
+	SSHHandshakeDeadline time.Duration
+	// BeforeHandle will be executed before `HandleSFTPSession` and before
+	// error checking happens during the socket listener.Accept().
+	BeforeHandle func(conn net.Conn, err error) bool
+	// HandleSFTPSession is executed when a new SFTP session is requested.
+	HandleSFTPSession func(channel ssh.Channel, sconn *ssh.ServerConn)
+}
+
+func defaultBeforeAccept(_ net.Conn, _ error) (accept bool) {
+	return true
+}
+
+// NewServer composes a new Server{} object from the options given.
+//
+// It is recommended to use (2*time.Minute) as the SSHHandshakeDeadline.
+// 2 minutes is the default deadline for OpenSSH servers/clients.
+func NewServer(options *Options) (sftpServer *Server, err error) {
+	if options.HandleSFTPSession == nil {
+		return nil, errors.New(missingConnectionHandlerFunction)
+	}
+	if options.SSHConfig == nil {
+		return nil, errors.New(missingSSHConfig)
+	}
+	if options.Logger == nil {
+		return nil, errors.New(missingLoggerInterface)
+	}
+	if options.Port < 1 || options.Port > 65535 {
+		return nil, errors.New(invalidPort)
+	}
+	if options.BeforeHandle == nil {
+		options.BeforeHandle = defaultBeforeAccept
+	}
+
+	listener, nErr := net.Listen(
+		"tcp",
+		net.JoinHostPort(options.PublicIP, strconv.Itoa(options.Port)),
+	)
+	if nErr != nil {
+		return nil, nErr
+	}
+
+	return &Server{
+		publicIP:             options.PublicIP,
+		port:                 options.Port,
+		sshConfig:            *options.SSHConfig,
+		sshHandshakeDeadline: options.SSHHandshakeDeadline,
+		handleSFTPSession:    options.HandleSFTPSession,
+		logger:               options.Logger,
+		listener:             listener,
+	}, nil
+}
+
+// Listen starts the SFTP server
+func (s *Server) Listen() (err error) {
+	s.logger.Info(ServerStarted,
+		fmt.Sprintf(
+			"SFTP Server listening on %s",
+			net.JoinHostPort(s.publicIP, strconv.Itoa(s.port)),
+		),
+	)
+
+	var tempDelay time.Duration // how long to sleep on accept failure
+	for {
+		conn, err := s.listener.Accept()
+		if !s.beforeHandle(conn, err) {
+			if conn != nil {
+				conn.Close()
+			}
+			continue
+		}
+		if err != nil {
+			// err.Temporary() has been depricated in favor of err.Timeout()
+			// However, the discussion around this seemed un-resolved.
+			// If we run into problems with this we need to investigate further.
+			// https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/10291/commits/a5cc7afaa718caf048d853576db8d5b03eb136ee
+			if ne, ok := err.(net.Error); ok && ne.Timeout() {
+				if tempDelay == 0 {
+					tempDelay = 5 * time.Millisecond
+				} else {
+					tempDelay *= 2
+				}
+				if max := 1 * time.Second; tempDelay > max {
+					tempDelay = max
+				}
+				s.logger.Error(AcceptNetworkError, fmt.Errorf("error while accepting connections: %w, retrying in %s", err, tempDelay))
+				time.Sleep(tempDelay)
+				continue
+			}
+			return err
```

Comment:
It seems a bit drastic to exit the entire server handler because of an error. We would the very least have to be sure this cannot be trigged by anything on the user side.

Overall, I think it would be better to just log and move on - otherwise it is a potential DOS target.

Reply:
Yeah. Didn't even consider the delay functionality, which basically can pretty much lock up incoming requests.

Reply:
Personally, I'd just take it all out. Log the error and move on.

Reply:
Agreed, I just did that. 

Currently I am distinguishing between Timeout() errors and non-timeout errors. With timeout errors triggering a continue and the non-timeout once triggering a full return. But honestly, i'm not sure about that either. I kind of feel like 99% of error cases here are going to merit a continue. 

I might look into this a bit more today.

Reply:
I looked into it and it seems that the Temporary functionality has not been changed, the method has just been marked as deprecated. 

I'm going to re-implement Temporary along with Timeout so that our new SFTP package is consistent with the current implementation.